File: ruby_solutions/uncompress.rb
def uncompress(s)
  result = ''
  i = 0

  while i < s.length
    num = ''
    while i < s.length && ('0'..'9').include?(s[i])
      num += s[i]
      i += 1
    end

    char = s[i]
    i += 1

    result += char * num.to_i
  end

  result
end

puts uncompress("2c3a1t") # -> 'ccaaat'
puts uncompress("4x2y1z") # -> 'xxxxyyz'
puts uncompress("3h5e2l1o") # -> 'hhheeeeeloo'
puts uncompress("1a1b1c1d") # -> 'abcd'
puts uncompress("10z") # -> 'zzzzzzzzzz'
puts uncompress("0m2n3o") # -> 'nnooo'

File: js_solutions/uncompress.js
const uncompress = (s) => {
  const validNums = '0123456789';
  let result = '';
  let i = 0;
  let j = 0;

  while (j < s.length) {
    if (validNums.includes(s[j])) {
      j++;
    } else {
      let num = s.slice(i, j)
      for (let count = 0; count < num; count++) {
        result += s[j];
      }
      j++;
      i = j;
    }
  }

  console.log(result)
}

uncompress("2c3a1t"); // -> 'ccaaat'
uncompress("4x2y1z"); // -> 'xxxxyyz'
uncompress("3h5e2l1o"); // -> 'hhheeeeeloo'
uncompress("1a1b1c1d"); // -> 'abcd'
uncompress("10z"); // -> 'zzzzzzzzzz'
uncompress("0m2n3o"); // -> 'nnooo'
uncompress("1a0b1c"); // -> 'ac'
uncompress("2p2q2r"); // -> 'ppqqrr'
uncompress("5a3b2c1d"); // -> 'aaaaabbbccd'
uncompress("1z"); // -> 'z'
uncompress("100x"); // -> 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
uncompress("0z1a"); // -> 'a'
